internal/drain-config/drainConfig.go

```diff
@@ -0,0 +1,200 @@
+package drain_config
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"slices"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/lib/pq"
+	"github.com/samber/lo"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	"github.com/rudderlabs/rudder-go-kit/logger"
+	migrator "github.com/rudderlabs/rudder-server/services/sql-migrator"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+const (
+	defaultPollFrequency      = 10
+	defaultPollFrequencyUnits = time.Second
+
+	defaultCleanupFrequency      = 1
+	defaultCleanupFrequencyUnits = time.Hour
+
+	defaultMaxAge      = 24
+	defaultMaxAgeUnits = time.Hour
+
+	// drain configurations
+
+	jobRunID       = "drain.jobRunIDs"
+	configJobRunID = "RSources.toAbortJobRunIDs"
+)
+
+type drainConfigManager struct {
+	log  logger.Logger
+	conf *config.Config
+	db   *sql.DB
+
+	done *atomic.Bool
+	wg   sync.WaitGroup
+}
+
+func NewDrainConfigManager(conf *config.Config, log logger.Logger) (*drainConfigManager, error) {
+	db, err := setupDBConn(conf)
+	if err != nil {
+		log.Errorw("db setup", "error", err)
+		return nil, fmt.Errorf("db setup: %v", err)
+	}
+	if err := migrate(db); err != nil {
+		log.Errorw("db migrations", "error", err)
+		return nil, fmt.Errorf("db migrations: %v", err)
+	}
+	return &drainConfigManager{
+		log:  log,
+		conf: conf,
+		db:   db,
+
+		done: &atomic.Bool{},
+	}, nil
+}
+
+func (d *drainConfigManager) CleanupRoutine(ctx context.Context) error {
+	d.wg.Add(1)
+	defer d.wg.Done()
+	for {
+		if d.done.Load() {
+			return nil
+		}
+		if _, err := d.db.ExecContext(
+			ctx,
+			"DELETE FROM drain_config WHERE created_at < $1",
+			time.Now().Add(-d.conf.GetDuration("drain.age", defaultMaxAge, defaultMaxAgeUnits)),
+		); err != nil && ctx.Err() == nil {
+			d.log.Errorw("db cleanup", "error", err)
+			return fmt.Errorf("db cleanup: %v", err)
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.After(
+			d.conf.GetDuration(
+				"drainConfig.cleanupFrequency",
+				defaultCleanupFrequency,
+				defaultCleanupFrequencyUnits,
+			),
+		):
+		}
+	}
+}
+
+func (d *drainConfigManager) DrainConfigRoutine(ctx context.Context) error {
+	d.wg.Add(1)
+	defer d.wg.Done()
+	// map to hold the config values
+	configMap := make(map[string][]string)
+	for {
+		if d.done.Load() {
+			return nil
+		}
+		// holds the config values fetched from the db
+		dbConfigMap := make(map[string][]string)
+
+		collectFromDB := func() error {
+			// fetch the config values from the db
+			rows, err := d.db.QueryContext(
+				ctx, "SELECT key, value FROM drain_config where key = ANY($1) and created_at > $2 ORDER BY key, value ASC",
+				pq.Array([]string{jobRunID}),
+				time.Now().Add(-1*d.conf.GetDuration("drain.age", defaultMaxAge, defaultMaxAgeUnits)),
+			)
```

Comment:
As long as we have control on the http api with respect to which keys we allow to be stored, we shouldn't need to have any key restrictions here